### PR TITLE
IoHandle: simplify interfaceName()

### PR DESCRIPTION
**Pull request by rgs1**
Was reading through #18531 and seeing the socket address being looked
up for each interface confused me a bit, so let's do it only once.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
cc: @junr03

File: source/common/network/io_socket_handle_impl.cc [hunk 1]
```diff
@@ -587,11 +587,19 @@ absl::optional<std::string> IoSocketHandleImpl::interfaceName() {
...
+      socket_address->ip()->version() != Address::IpVersion::v6) {
+    ENVOY_BUG(false, fmt::format("unexpected IP family {}", socket_address->ip()->version()));
+  }
+
   Api::InterfaceAddressVector interface_addresses{};
   const Api::SysCallIntResult rc = os_syscalls_singleton.getifaddrs(interface_addresses);
   RELEASE_ASSERT(!rc.return_value_, fmt::format("getiffaddrs error: {}", rc.errno_));
 
   absl::optional<std::string> selected_interface_name{};
+  absl::uint128 socket_address_value = socket_address->ip()->version() == Address::IpVersion::v4
```
Comment (alyssawilk) [inline source/common/network/io_socket_handle_impl.cc:600]: I think the prior version would return an empty interface name but I think falling through here might segfault instead.  Maybe add a unit test?

**Comment by junr03** [comment]
@rgs1 are you planning on updating with Alyssa's comment?

**Comment by alyssawilk** [comment]
I think he's out on leave right now, so I assume it'll wait until he gets back

**Pull request closed without merging**